File: problems/栈与队列/20.有效的括号.cpp
#include <iostream>
#include <stack>
#include <unordered_map>
using namespace std;

bool isValid(string s) {
    stack<char> st;
    unordered_map<char, char> mapping = {{')','('}, {'}','{'}, {']','['}};
    for (char c : s) {
        if (mapping.count(c)) {
            char top = st.empty() ? '#' : st.top();
            if (mapping[c] != top) return false;
            if (!st.empty()) st.pop();
        } else {
            st.push(c);
        }
    }
    return st.empty();
}

int main() {
    // 测试用例
    cout << (is_valid("()") ? "True" : "False") << endl;      // 预期: True
    cout << (is_valid("()[]{}") ? "True" : "False") << endl;  // 预期: True
    cout << (is_valid("(]") ? "True" : "False") << endl;      // 预期: False
    cout << (is_valid("([)]") ? "True" : "False") << endl;   // 预期: False

    return 0;
}

File: problems/栈与队列/20.有效的括号.py
#!/usr/bin/env python3
"""
20. 有效的括号
算法: 栈
时间复杂度: O(n)
空间复杂度: O(n)
"""
def is_valid(s: str) -> bool:
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}
    for ch in s:
        if ch in mapping:
            top = stack.pop() if stack else '#'
            if mapping[ch] != top:
                return False
        else:
            stack.append(ch)
    return not stack


if __name__ == "__main__":
    print(is_valid("()"))      # 预期: True
    print(is_valid("()[]{}"))  # 预期: True
    print(is_valid("(]"))      # 预期: False
    print(is_valid("([)]"))     # 预期: False
